app/models/display_ad.rb

```diff
@@ -32,9 +32,20 @@ class DisplayAd < ApplicationRecord
                              search: "%#{term}%"
                      }
 
-  def self.for_display(area, user_signed_in)
+  def self.for_display(area, user_signed_in, article_tags = [])
     relation = approved_and_published.where(placement_area: area).order(success_rate: :desc)
 
+    if article_tags.any?
```

Comment:
This all looks logically correct to me, and is backed by a solid amount of tests by the look of it.

In anticipation of further complexity being added to this area, I'd say we should be okay to ship this right now, while also planning to maybe extract this whole method into a dedicated object and then break the steps into sub-methods with super expressive names to describe the steps and hide the mess.

I don't think this needs to be a blocker since we anticipate continuing work in this area, so not a lot of risk of moving on, I'd just say let's keep this refactor in mind to allow this method to be maximally understood from fresh eyes seeing it down the road, and able to accept future changes which could bring inherent complexity.

That's all! This strikes me as totally good for now, but looks like it's ready to be refactored as part of whatever the next step is. 😄

Reply:
Thank Ben, yip we're definitely on the same page about extracting this out as a next step!